### PR TITLE
[GOLD] Cherry-pick status pointer to reference replacement from master

### PR DESCRIPTION
**Pull request by rlnx**
See #1083

### CODE REVIEW

**Review by napetrov — APPROVED** [review]

File: cpp/daal/include/data_management/data/internal/numeric_table_sycl_homogen.h [hunk 1]
```diff
@@ -338,6 +343,7 @@ class DAAL_EXPORT SyclHomogenNumericTable : public SyclNumericTable
         static services::Status read(const services::internal::Buffer<U> & buffer, BlockDescriptor<T> & block, size_t nRows, size_t nCols)
         {
             DAAL_ASSERT(buffer.size() == nRows * nCols);
```
Comment (mchernov-intel) [inline cpp/daal/include/data_management/data/internal/numeric_table_sycl_homogen.h:345]: check multiplication via `DAAL_OVERFLOW_CHECK_BY_MULTIPLICATION`

File: cpp/daal/include/data_management/data/internal/numeric_table_sycl_homogen.h [hunk 2]
```diff
@@ -437,7 +457,10 @@ class DAAL_EXPORT SyclHomogenNumericTable : public SyclNumericTable
             const size_t nRows     = block.getNumberOfRows();
             const size_t rowOffset = block.getRowsOffset();
 
-            status |= BufferIO<T, DataType>::write(getSubBuffer(rowOffset, nRows), block, nRows, nCols);
+            const services::internal::Buffer<DataType> subBuffer = getSubBuffer(rowOffset, nRows, status);
+            DAAL_CHECK_STATUS_VAR(status);
+
+            status |= BufferIO<T, DataType>::write(subBuffer, block, nRows, nCols);
```
Comment (mchernov-intel) [inline cpp/daal/include/data_management/data/internal/numeric_table_sycl_homogen.h:463]: add `DAAL_CHECK_STATUS_VAR(status);`

File: cpp/daal/include/data_management/data/internal/numeric_table_sycl_soa.h [hunk 3]
```diff
@@ -460,7 +467,8 @@ class DAAL_EXPORT SyclSOANumericTable : public SyclNumericTable
         if (onDeserialize)
         {
             rwMode = readWrite;
-            allocateDataMemoryImpl();
+            st |= allocateDataMemoryImpl();
+            services::throwIfPossible(st);
```
Comment (mchernov-intel) [inline cpp/daal/include/data_management/data/internal/numeric_table_sycl_soa.h:471]: add `DAAL_CHECK_STATUS_VAR(st);`

File: cpp/daal/include/services/internal/buffer.h [hunk 4]
```diff
@@ -120,50 +120,93 @@ class Buffer : public Base
...
+    /**
+     *  Converts data inside the buffer to the host side, throws exception if conversion fails
+     *  \param[in]  rwFlag  Access flag to the data
+     *  \return host-allocated shared pointer to the data
+     */
+    SharedPtr<T> toHost(const data_management::ReadWriteMode & rwFlag) const
+    {
+        services::Status status;
+        const SharedPtr<T> ptr = toHost(rwFlag, status);
+        services::throwIfPossible(status);
```
Comment (mchernov-intel) [inline cpp/daal/include/services/internal/buffer.h:143]: Does throw is possible everywhere? What if no?

File: cpp/daal/include/services/internal/buffer.h [hunk 5]
```diff
@@ -120,50 +120,93 @@ class Buffer : public Base
...
+     *  \param[out] status  Status of operation
      *  \return one-dimensional SYCL* buffer
      */
-    cl::sycl::buffer<T, 1> toSycl(Status * status = NULL) const
+    cl::sycl::buffer<T, 1> toSycl(Status & status) const
     {
         if (!_impl)
         {
-            internal::tryAssignStatusAndThrow(status, ErrorEmptyBuffer);
+            status |= ErrorEmptyBuffer;
```
Comment (mchernov-intel) [inline cpp/daal/include/services/internal/buffer.h:158]: Add `services::throwIfPossible(status);`

**Pull request merged**